WIP: Convert tplink integration to use a new asyncio-enabled backend library

rytilahti opened the pull request:
## Description:
This PR ports the tplink integration to use a maintainer-made asyncio-fork of pyhs100 called python-kasa.

The major changes which should help with the linked issues are:
* Uses asyncio to avoid blocking I/O.
* The responses are cached now, the data is fetched from devices by awaiting `update()`.
* Only the main device is polled for smartstrips, dropping number of I/O requests for multiplug devices, thanks to an anonymous donor for sending me a test device! :tada:
* All data is fetched using a single I/O operation. This will effectively drop the number of requests per cycle to 1 per device.
* Bulbs and dimmers now support defining transition periods for changes.
* Support for light strips (KL430).

This PR has been tested to work with the following devices so far:
* HS110
* KP303
* KL60
* KL130

**It would be great to get feedback from users with other devices!**
Testing this requires installing python-kasa from git: https://github.com/python-kasa/python-kasa/ .

**Related issue (if applicable):** Related to / potentially closes: #30357, closes #30693, closes #28912, closes #23045, closes #28590, closes #29169, obsoletes #35628, obsoletes #35460, obsoletes #34465, obsoletes #29941

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ ] There is no commented out code in this PR.
  - [ ] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

rytilahti committed:
Initial port of tplink integration over to python-kasa

balloob commented on homeassistant/components/tplink/light.py:
```diff
@@ -55,12 +55,12 @@
     # exception that is caught by async_add_entities_retry which
     # will try again later.
-    device.get_sysinfo()
+    await device.update()
```
Here you're updating it again, but you also just updated it during discovery ?

rytilahti replied:
I tried to keep the existing semantics in place for the initial testing. The reasoning for this update was to trigger an exception during the initial setup (maybe the device was manually configured, but not available during the startup?).

balloob commented on homeassistant/components/tplink/light.py:
```diff
@@ -212,27 +212,18 @@ def is_on(self):
+                self._light_features or self.get_light_features()
             )
-            self._light_state = self.get_light_state_retry(self._light_features)
+            self._light_state = self.get_light_state()
```
This should either be `await` or you're doing I/O in async context.

balloob commented on homeassistant/components/tplink/light.py:
```diff
@@ -212,27 +212,18 @@ def is_on(self):
             # Update light features only once.
             self._light_features = (
-                self._light_features or self.get_light_features_retry()
+                self._light_features or self.get_light_features()
```
If `self._light_features` is 0, this is executed each time. Probably better check for `None`.

rytilahti replied:
I think this should never come back as 0, as all the bulbs support one of those features.

balloob commented on homeassistant/components/tplink/light.py:
```diff
@@ -355,47 +322,39 @@ def get_light_state(self, light_features: LightFeatures) -> LightState:
+
 
+        _LOGGER.warning("Could not set state for %s", self.smartbulb.host)
         self._is_setting_light_state = False
```
You can put this in a `finally` as part of the `try…except`

rytilahti replied:
Finally would be executed also in the case where setting the state succeeded after a retry.

balloob replied:
Right, which it also sets it to False ?

rytilahti replied:
There is a return inside the try-block, so in case of success that warning will never be hit.

MartinHjelmare replied:
Just move the setting of `_is_setting_light_state` to `False` to finally, not the warning.

rytilahti replied:
That will work after removing the retry logic, otherwise it would still require some extra logic to avoid setting it to false if there are still retries left.

balloob commented on homeassistant/components/tplink/switch.py:
```diff
@@ -88,6 +88,7 @@ def device_info(self):
             "name": self._alias,
             "model": self._model,
             "manufacturer": "TP-Link",
+            # FIXME: should this be the MAC even for subdevices?
```
Yeah. I would group them all in 1 device as that's what it is

balloob commented on homeassistant/components/tplink/switch.py:
```diff
@@ -102,46 +103,34 @@ def is_on(self):
-                self._state = self.smartplug.state == self.smartplug.SWITCH_STATE_ON
-            else:
-                self._state = self._plug_from_context["state"] == 1
+                self._alias = self.smartplug.alias
```
Why set all these as instance variables on the class and not have `name` property return `self.smartplug.alias` ?

rytilahti replied:
These are leftovers and are not anymore necessary as the caching is done in the backend. I'll remove the local variables.

One open question here, how to trigger re-read for `device_info` property, in case the device name changes?

balloob replied:
Call `device_registry.async_update_device` with the new name.

rytilahti replied:
Should that be called always under `async_update` (i.e., is there some harm by doing that), or should I restore caching for that specific case?

rytilahti committed:
Add missing await for state getting

rytilahti committed:
Some cleanups to switch based on review

* Remove unnecessary local caching, as this is done in the backend
* Request update after turn_on and turn_off (the backend does update the values on these already)

rytilahti committed:
Fix linting

vangorra commented:
You're removing the retry logic for the light update. Why is that? The new kasa api doesn't seem to support it. The retry is there to avoid common issues with the bulb throwing an error instead of providing data when it gets busy. The new API will cause more calls to the bulb as it refreshes sysinfo on every update call.

rytilahti commented:
I removed it for the time being, as simply doing a single retry by calling another method does not necessitate the extra code branching (it just made it more difficult to follow what is going on); if retries are wanted, they should be done either by a loop or some other structure.

The main point for doing this lib cleanup is to fix the underlying issues causing the need for so many requests, so my hoping is that in the end there will be no need for such retry behavior. And if it is still needed, I'll simply add it to the backend library, it should be invisible to homeassistant what is happening underneath the hood.

vangorra commented:
Adding retries to the back end library is the best approach. However removing it entirely will result in failed lighting commands. This isn't a problem with the tplink plugs. I suspect the web server on the lights can only handle serial requests with spacing between them.

rytilahti commented:
I had some bulbs for testing last year and that's when I implemented caching in the backend library. The problem was way too many requests on them in a short period of time, mostly accessing the information that's already available after receiving the initial system information from them.

But I'll keep that in mind when I get so far, in its current state this should already work just fine with the bulbs though.

dmcavinue commented:
I have a couple of HS300s at home that have this issue which I have just been ignoring. I'll spin up this change in my deployment later on and see if it resolves my communication errors and drop some feedback.

rytilahti commented:
If you want to give it a try, you should use the code from [this PR](https://github.com/python-kasa/python-kasa/pull/26) or stay tuned until it gets merged. This integration PR in its current state should work well enough for testing, but some changes are needed to adapt the "subdevices" to be non-polling.

uncled1023 commented:
Any updates on this PR?  After needing to perform a clean install, i'm unable to use my TP-Link devices because they can't be discovered.

rytilahti committed:
remove deprecated async_setup_platform to synchronize with dev

rytilahti committed:
Add proper support for multisocket devices (poll only the parent)

* special thanks to anonymous donater of kp303!

* there is no need to poll subdevices, let the parent inform them on updates
  * this alone will drop the number of necessary requests from <number of sockets>+1 to 1 (for the sysinfo).
* decreased MIN_TIME_BETWEEN_UPDATES to 5
* handle smartdimmer as dimmers

balloob commented on homeassistant/components/tplink/light.py:
```diff
@@ -163,42 +130,45 @@ def available(self) -> bool:
-        await self._async_set_light_state_retry(
+        brightness = (
+            int(kwargs[ATTR_BRIGHTNESS])
+            if ATTR_BRIGHTNESS in kwargs
+            else self._light_state.brightness
+            if self._light_state.brightness is not None
+            else 255
```
We shouldn't use inline if-else statements if they wrap multiple lines. It hurts readability.

rytilahti replied:
Yes, my plan is to merge the relevant commits from #33606 to this PR, which will also fix that. I'm just waiting for a test bulb to arrive, so that I can verify that things still work after I do that.

rytilahti committed:
refactor to avoid unnecessary code & optimize device accesses

the backend lib will after https://github.com/python-kasa/python-kasa/pull/59
fetch all the necessary wanted in a single update cycle.

rytilahti committed:
remove retries, we are updating every fives seconds anyway

rytilahti committed:
make sure the device supports feature before trying to access it, reset setting state always

rytilahti committed:
cast hs to int, define helper vars only where needed

rytilahti committed:
remove double-conversion for brightness

rytilahti committed:
convert brightness for brightness-only mode, use device props for supported checks

vangorra commented:
How can we test this code? `kasa` is not in pypi and running `pip install git+https://github.com/python-kasa/python-kasa.git` fails.

bdraco commented:
> How can we test this code? `kasa` is not in pypi and running `pip install git+https://github.com/python-kasa/python-kasa.git` fails.

I tested with with cloning `python-kasa` and then running `pip3 install --upgrade .`

vangorra commented:
I managed to get `python-kasa` to build independent of hass. Getting it to install into hass is proving problematic.
```
$ pip install ~/Documents/git/python-kasa/
Processing /home/rvangork/Documents/git/python-kasa
  Installing build dependencies ... error
  ERROR: Command errored out with exit status 1:
   command: /home/rvangork/Documents/git/home-assistant/venv/bin/python3 /home/rvangork/Documents/git/home-assistant/venv/lib/python3.8/site-packages/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-z21_aybn/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'poetry>=0.12'
       cwd: None
  Complete output (45 lines):
  Keyring is skipped due to an exception: 'PosixPath' object has no attribute 'read_text'
  Traceback (most recent call last):
    File "/usr/lib/python3.8/runpy.py", line 193, in _run_modu
```

rytilahti committed:
cleanup light, no need to check if we are already setting the state

rytilahti committed:
cleanup switch: try-except only the parts doing I/O, be optimistic on child plug state changes

rytilahti commented:
@vangorra sorry, I didn't see your comment until now..

You can simply clone the repository and command `poetry install` which takes care of everything. If you have a virtualenv active, it will be installed inside that env. Otherwise, a new virtualenv will be automatically created (`poetry shell` can be used to open the shell, `poetry run kasa` to run the console tool directly).

vangorra commented:
Code is running. Most of my tplink devices fail to add to hass because of the following in update_emeter_state:
```
2020-05-28 06:20:54 ERROR (MainThread) [homeassistant.components.light] tplink: Error on device update!
Traceback (most recent call last):
  File "/home/rvangork/Documents/git/home-assistant/homeassistant/helpers/entity_platform.py", line 322, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/home/rvangork/Documents/git/home-assistant/homeassistant/helpers/entity.py", line 470, in async_device_update
    await self.async_update()
  File "/home/rvangork/Documents/git/home-assistant/homeassistant/components/tplink/light.py", line 190, in async_update
    self.update_emeter_state()
  File "/home/rvangork/Documents/git/home-assistant/homeassistant/components/tplink/light.py", line 231, in update_emeter_state
    self.smartbulb.emeter_realtime["power"]
AttributeError: 'SmartBulb' object has no attribute 'emeter_realtime'
```

rytilahti commented:
For the first error, it looks like you are running too old git version of python-kasa, the property should be defined in smartdevice.py of your checkout like this: https://github.com/python-kasa/python-kasa/blob/master/kasa/smartdevice.py#L335

For the second one, I haven't touched the existing retry code of those parts but that should be fixed. I don't think there is a need to block the start if already registered (as in, available in device registry) devices are offline.

darthbodacious commented:
Hello! I've been following this issue (and the many that have found their way to be linked here) because I went all in on HS300 strips for a new setup.

While I know I have a clear version of the problem that this PR is looking to fix, I'm unsure if I'm also a suitable tester (which I know folks above have seemed to work out).

I'm running Home Assistant on a RPi3b+... is there a way for me to try this code out? Or would I need to be running another install method to realistically contribute?

darthbodacious commented:
Oh! I figured it out (thanks to the comments here and the documentation!)
I set up a separate `venv` instance on my laptop, installed the `python-kasa` master branch and the `home-assistant` branch from this PR using `pip3 install --upgrade .` in that virtual environment... then got up and running / copied over my config files from the pi to test things out.

First time I booted it up, all 13 HS300 strips and 5 HS105 plugs integrated flawlessly (and I ended up with 96 entries total, since it looks like the strips themselves now self-register as entries).

Hopefully that's as good of news to you as it was to me!

rytilahti commented:
@darthbodacious hi and thanks for testing, that's greatly appreciated! Please let me know how/if it still works after using it a while :-) 

As the last big upstream PRs got merged lately, I'm planning to do a pip-installable (pre-)release soonish when I'll find some free time. That will make it much easier to test this PR (by using the files from this PR as a custom_component), which will allow us to squash the last remaining bugs before getting this finally merged.

There is one issue that should probably be fixed somehow before the merge (https://github.com/python-kasa/python-kasa/issues/64), but I'm unsure how to approach that. One the one hand, I'm pretty sure some people want to use the emeter information on socket basis. On the other hand, starting to request that information from each socket separately (meaning raising the number of requests from 1 to 7 per update cycle) may potentially reintroduce the connectivity issues this PR is aiming to fix. 

One potential fix would be to make that configurable, if there is no other way to handle it. Another possibility that is worth exploring is the fact that the the API parameter used to declare which subdevices to act on is literally called `child_ids`, which takes a list as input. Maybe this could allow to fetch updates from all the sockets in a single request (and keep on using a single request for all data)?

rytilahti committed:
depend on python-kasa's pre-release to allow easier testing

vangorra commented:
Using the latest `python-kasa` works. This is still blocking startup. Home assistant places this snackbar on the ui `Home Assistant is starting, not everything will be available until it is finished.`

```
2020-06-07 20:29:51 DEBUG (MainThread) [homeassistant.components.tplink.common] Attempting to add object of type <class 'kasa.smartbulb.SmartBulb'>
2020-06-07 20:29:51 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.tplink entity: light.office_lamp
2020-06-07 20:29:52 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.tplink entity: light.office_east
2020-06-07 20:29:52 DEBUG (MainThread) [homeassistant.components.tplink.common] Added object.
2020-06-07 20:29:52 DEBUG (MainThread) [homeassistant.components.tplink.common] Attempting to add object of type <class 'kasa.smartbulb.SmartBulb'>
2020-06-07 20:29:52 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.tplink entity: light.child1_closest
2020-06-07 20:29:52 DEBUG (MainThread) [homeassistant.components.tplink.common] Added object.
2020-06-07 20:29:52 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new light.tplink entity: light.child1_room
2020-06-07 20:29:56 DEBUG (MainThread) [homeassistant.components.tplink.common] Unable to query the device: [Errno 113] Connect call failed ('10.40.12.93', 9999)
2020-06-07 20:29:56 DEBUG (MainThread) [homeassistant.components.tplink.common] Failed to add object, will try again later
2020-06-07 20:30:20 ERROR (MainThread) [homeassistant.components.switch] Setup of platform tplink is taking longer than 60 seconds. Startup will proceed without waiting any longer.
2020-06-07 20:30:20 ERROR (MainThread) [homeassistant.components.light] Setup of platform tplink is taking longer than 60 seconds. Startup will proceed without waiting any longer.
2020-06-07 20:30:20 INFO (MainThread) [homeassistant.bootstrap] Home Assistant initialized in 66.02s
2020-06-07 20:30:20 INFO (MainThread) [homeassistant.core] Starting Home Assistant
```

rytilahti commented:
Maybe that has something to do with the retry functionality? The backend library has timeouts for request (and it does 3 retries per default when request fails, but this should not be blocking).

darthbodacious commented:
> @darthbodacious hi and thanks for testing, that's greatly appreciated! Please let me know how/if it still works after using it a while :-)
> 
> As the last big upstream PRs got merged lately, I'm planning to do a pip-installable (pre-)release soonish when I'll find some free time. That will make it much easier to test this PR (by using the files from this PR as a custom_component), which will allow us to squash the last remaining bugs before getting this finally merged.
> 
> There is one issue that should probably be fixed somehow before the merge ([python-kasa/python-kasa#64](https://github.com/python-kasa/python-kasa/issues/64)), but I'm unsure how to approach that. One the one hand, I'm pretty sure some people want to use the emeter information on socket basis. On the other hand, starting to request that information from each socket separately (meaning raising the number of requests from 1 to 7 per update cycle) may potentially reintroduce the connectivity issues this PR is aiming to fix.
> 
> One potential fix would be to make that configurable, if there is no other way to handle it. Another possibility that is worth exploring is the fact that the the API parameter used to declare which subdevices to act on is literally called `child_ids`, which takes a list as input. Maybe this could allow to fetch updates from all the sockets in a single request (and keep on using a single request for all data)?

Since my primary instance of home assistant runs on the Pi rather than the laptop that I used for testing... I haven't "used" this branch much more since commenting. But I'll give it another spin to see if anything jumps out to me. And I'd love to help out, generally, any way that I can!

I don't have much insight into the questions you're posing here, but I'd be happy to try running anything / gathering any data that might help!

draco2003 commented:
@rytilahti  Saw this in the node module that interfaces with tplink hardware incase it helps you with the child_ids logic
https://github.com/plasticrake/tplink-smarthome-api/blob/c8f8aa61ec2b1e61fb5aaba4adecfc5b72504a1f/src/device/index.ts#L425

rytilahti commented:
Hi @draco2003 and thanks for the hint! Looks like the logic is the same that is being used in python-kasa, so I think I'll go with the plan described here: https://github.com/home-assistant/core/issues/28912#issuecomment-651955108

rytilahti committed:
Add support for bulb transitions

vangorra commented:
@rytilahti I copied the code from #34465 into this PR and it fixed the blocking on startup as well. Here is the common file.
```python
"""Common code for tplink."""
import asyncio
from dataclasses import dataclass
from datetime import timedelta
import logging
from typing import Any, Awaitable, Callable, List

from kasa import (
    Discover,
    SmartBulb,
    SmartDevice,
    SmartDeviceException,
    SmartDimmer,
    SmartPlug,
    SmartStrip,
)

from homeassistant.core import callback
from homeassistant.helpers.event import async_track_time_interval
from homeassistant.helpers.typing import HomeAssistantType

_LOGGER = logging.getLogger(__name__)


ATTR_CONFIG = "config"
CONF_DIMMER = "dimmer"
CONF_DISCOVERY = "discovery"
CONF_LIGHT = "light"
CONF_STRIP = "strip"
CONF_SWITCH = "switch"


class SmartDevices:
    """Hold different kinds of devices."""

    def __init__(
        self, lights: List[SmartDevice] = None, switches: List[SmartDevice] = None
    ):
        """Initialize device holder."""
        self._lights = lights or []
        self._switches = switches or []

    @property
    def lights(self):
        """Get the lights."""
        return self._lights

    @property
    def switches(self):
        """Get the switches."""
        return self._switches

    def has_device_with_host(self, host):
        """Check if a devices exists with a specific host."""
        for device in self.lights + self.switches:
            if device.host == host:
                return True

        return False


async def async_get_discoverable_devices(hass):
    """Return if there are devices that can be discovered."""
    return await Discover.discover()


async def async_discover_devices(
    hass: HomeAssistantType, existing_devices: SmartDevices
) -> SmartDevices:
    """Get devices through discovery."""
    _LOGGER.debug("Discovering devices")
    devices = await async_get_discoverable_devices(hass)
    _LOGGER.info("Discovered %s TP-Link smart home device(s)", len(devices))

    lights = []
    switches = []

    for dev in devices.values():
        await dev.update()
        # If this device already exists, ignore dynamic setup.
        if existing_devices.has_device_with_host(dev.host):
            continue

        if dev.is_strip or dev.is_plug:
            switches.append(dev)
        elif dev.is_dimmer or dev.is_bulb:
            lights.append(dev)
        else:
            _LOGGER.error("Unknown smart device type: %s", dev.device_type)

    return SmartDevices(lights, switches)


def get_static_devices(config_data) -> SmartDevices:
    """Get statically defined devices in the config."""
    _LOGGER.debug("Getting static devices")
    lights = []
    switches = []

    for type_ in [CONF_LIGHT, CONF_SWITCH, CONF_STRIP, CONF_DIMMER]:
        for entry in config_data[type_]:
            host = entry["host"]

            if type_ == CONF_LIGHT:
                lights.append(SmartBulb(host))
            elif type_ == CONF_SWITCH:
                switches.append(SmartPlug(host))
            elif type_ == CONF_STRIP:
                switches.append(SmartStrip(host))
            elif type_ == CONF_DIMMER:
                lights.append(SmartDimmer(host))

    return SmartDevices(lights, switches)


@dataclass
class ProcessObjectResult:
    """The result of processing an object in async_add_entities_retry."""

    object: Any
    result: bool


AsyncAddEntities = Callable[[List[Any], bool], None]


async def async_add_entities_retry(
    hass: HomeAssistantType,
    async_add_entities: AsyncAddEntities,
    objects: List[Any],
    add_entity_callback: Callable[[Any, AsyncAddEntities], Awaitable],
    interval: timedelta = timedelta(seconds=60),
) -> Callable[[], None]:
    """
    Add entities now and retry later if issues are encountered.

    If the callback throws an exception or returns false, that
    object will try again a while later.
    This is useful for devices that are not online when hass starts.
    :param hass:
    :param async_add_entities: The callback provided to a
    platform's async_setup.
    :param objects: The objects to create as entities.
    :param add_entity_callback: The callback that will perform the add.
    :param interval: THe time between attempts to add.
    :return: A callback to cancel the retries.
    """
    add_objects = list(objects)
    cancel_func1 = None

    @callback
    def cancel() -> None:
        nonlocal cancel_func1
        cancel_func1()

    async def process_object(add_object) -> ProcessObjectResult:
        nonlocal add_entity_callback
        nonlocal async_add_entities
        try:
            _LOGGER.debug("Attempting to add object of type %s", type(add_object))
            result = await add_entity_callback(add_object, async_add_entities)
            if result is not False:
                _LOGGER.debug("Object added")
                result = True

        except SmartDeviceException as ex:
            _LOGGER.debug(str(ex))
            result = False

        return ProcessObjectResult(object=add_object, result=result)

    @callback
    async def process_objects(*args) -> None:
        nonlocal add_objects

        # Process objects in parallel.
        results: List[ProcessObjectResult] = await asyncio.gather(
            *[process_object(add_object) for add_object in tuple(add_objects)],
            loop=hass.loop,
        )

        for result in results:
            if result.result:
                add_objects.remove(result.object)

        if add_objects:
            _LOGGER.debug("Waiting to try adding objects again.")
        else:
            cancel()

    # Schedule the retry interval.
    cancel_func1 = async_track_time_interval(hass, process_objects, interval)

    # Attempt to add the objects.
    await process_objects()

    return cancel

```

vangorra commented:
I noticed that most the tests fail because SmartBulb doesn't have an updated method. I pulled the latest python-kasa and installed it in my venv but still have the result.

rytilahti commented:
@vangorra could you create a PR based on this PR for the blocking fix? Looks like the tests are failing as they haven't been converted, I'll look into that

vangorra committed:
Fixing bug whre tplink potentially blocked HASS from ever starting.

vangorra commented:
PR for startup blocking fix created: https://github.com/rytilahti/home-assistant/pull/2

rytilahti committed:
pyhs100 -> kasa for tests

balloob commented on homeassistant/components/tplink/manifest.json:
```diff
@@ -3,6 +3,6 @@
   "name": "TP-Link Kasa Smart",
   "config_flow": true,
   "documentation": "https://www.home-assistant.io/integrations/tplink",
-  "requirements": ["pyHS100==0.3.5"],
```
I really want to get pyHS100 out of Home Assistant as it is the last dependency that depends on `typing` and so prevents us from blocking `typing` from being installed, something that breaks new venv installations and CI unless we put extra code to fix it.

balloob replied:
This has been resolved.

draco2003 commented:
@rytilahti appreciate all of the work on this PR! 
What's left to be done prior to merging?  
Is there anything we can help you with on this?

rytilahti commented:
@draco2003 hey, thanks for chiming in! There are two roadblockers at the moment:
* HS300 emeter support: this will require some backend changes to make it properly, I'm currently waiting for access to a device that was gratefully offered by a user.
* Unit tests need to be adapted to the changes, I haven't yet looked into those due to the first point.
* Get some people to test it, I want to avoid breaking existing installations even when a breaking change is already coming up (to create separate entities for sensors).

@vangorra thanks for the PR, I'll take a look at it and do a merge this weekend!

draco2003 commented:
@rytilahti I know you are working on getting connected to a remote device, if that becomes impractical, I'd be happy to send or fund an HS300 + converter so that you're able to test it as needed. Keep me posted there. 

I'm happy to help with the unit test work, once the backend changes are done. 

If you can point me to an easy way to test using the hassio install on RPI I'd be happy to test the current version as well.

rytilahti committed:
Allow adding lightstrips based on static config or discovery

rosterloh commented:
I couldn't communicate with my devices since 0.112 but I've integrated this PR and it fixes it for me. Great work!

notalecdusheck commented:
Many people are having issues with the Kasa integration and this seems to fix everything. id like to see this PR merged

TyGooch commented:
Thank you @rytilahti for all time you've put into this! I've had connectivity issues with 2 HS300s for months and this PR did wonders for my setup.

Currently, emeter info is only working for one plug on each device but it's a trade off that I'd say is worth it for anyone having issues with an HS300. No more unavailable plugs and unresponsive switches 🥳

sjuxax commented:
Almost works for me -- requires a bump to python-kasa==0.4.0.dev1 to avoid failure to import `SmartLightStrip`.

morninglite commented:
> Currently, emeter info is only working for one plug on each device but it's a trade off that I'd say is worth it for anyone having issues with an HS300. 

The only purpose of my HS300 is to monitor energy consumption, so I would strongly disagree.

I look forward to to the new python-kasa as well, but would appreciate it not being merged before that is fixed.

AlecDusheck commented:
I’d personally rather have this integration working with a single feature missing rather then having the Kasa integration not working at all.

scottismyname commented:
Agreed, this integration is unusable for me right now. If you don't want to
change, don't upgrade. Just my opinion

On Sun, Aug 9, 2020, 3:55 PM Alec Dusheck <notifications@github.com> wrote:

> I’d personally rather have this integration working with a single feature
> missing rather then having the Kasa integration not working at all.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/30719#issuecomment-671110745>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGYSVMKHGNUIMYIQQJOE5TR74SORANCNFSM4KF2RNBQ>
> .
>

morninglite commented:
Sure, easy for you to say, when you don't use that feature.

For the current integration there's at least a [workaround available](https://github.com/home-assistant/core/pull/29941) that doesn't take away any functionality. It might be dirty, but it's been proven to work for myself, as well as others, for over 6 months.

connorproctor commented:
There a new feature (HS220 transition support) in python-kasa that I would love to get integrated with home assistant, but that can't happen until this PR is merged.

loaxley commented:
Thanks everyone for working on this...looking forward to when it gets merged.

MarkHofmann11 commented:
> Sure, easy for you to say, when you don't use that feature.
> 
> For the current integration there's at least a [workaround available](https://github.com/home-assistant/core/pull/29941) that doesn't take away any functionality. It might be dirty, but it's been proven to work for myself, as well as others, for over 6 months.

Agreed - While we all are excited about this major new update, this needs to work 100% and have ample testing.  There has been a work-around for people with tplink issues for over 6 months.  I have been running the modified tplink work-around code ever since December of 2019 and it works 100%. with all my tplink devices including the KP400 with multiple outlets.  I would have liked to have the work-around merged, but it was never approved.

rytilahti commented:
Sorry for the late response, so first of all, this is not going to be merged before it has the feature parity to the current implementation (HS300 emeter readings), so no worries about that.

I was previously against merging any "hacks" to fix some cases to avoid making it harder to keep this PR in sync, but considering it remains unclear when proper HS300 support lands to python-kasa, I think the way to go is to merge the fixes as long as someone takes over the ownership (or acts as second codeowner for tplink) as I don't currently have interest nor time to keep fixing it in case something breaks.

Considering this is almost a rewrite of the integration, I think the safest approach is to take the current state of this PR and incorporate it inside python-kasa as `tplinkng` (or similar) custom integration. This will allow changes to the current integration while also making it easier for anyone interested in using the python-kasa based version (even alongside the current one, for those who want new features not currently available and still want to have HS300 emeter support available). There are also some other changes I would like to have, e.g., separating sensors to their own entities as they should be.

When the time comes that the new integration is mature enough to be merged back to the mainline, I'll create a PR to replace the old implementation. As there is no ETA, who knows when that's going to happen, but I think two versions living side-by-side for now would be a workable compromise to fix the current unfortunate situation.

What do you think about such a solution? I proposed @vangorra that he'd take up the maintainership (or at least share it with me) for the time being, but haven't heard back from him yet. If someone else wants to step up to maintain the integration, I'm open for suggestions.

@loaxley that would be great as I haven't unfortunately been able to obtain access so far. The promised access didn't pan out (yet), life happens :-)

MarkHofmann11 commented:
@rytilahti I think what you are suggesting is the best approach.  @TheGardenMonkey gets the credit for the work-around that I started using back in December 2019.  I had recently updated his modification to include the recent updates since that date - here:  https://github.com/home-assistant/core/pull/35460  I have a bunch of different tplink devices, and lots of them - all working perfectly using this work-around.  Still running this over here currently with no issues.

TheGardenMonkey commented:
I am still down for assisting with this integration. I have a large number of the tp-link / kasa devices (added even more since I wrote the other PR).

mkarnebeek commented:
This is still relevant. My tplink plugs require me to restart multiple times before they all get picked up

scottismyname commented:
I have given up on the integration for now, it's too unstable. Luckily I can integrate my plugs with smartthings and use them that way. Looking forward to a more stable release so I can switch back. I'd prefer to avoid using st for things that don't need them

TheGardenMonkey commented:
> This is still relevant. My tplink plugs require me to restart multiple times before they all get picked up

Can you try out the one from my PR which is still using the old library until the new one is ready...https://github.com/home-assistant/core/pull/39762 | https://github.com/TheGardenMonkey/core/tree/dev/homeassistant/components/tplink. I have been using this logic since last year without issues but would like to get some more field testing.

rytilahti commented:
I'll close this now, see https://github.com/home-assistant/core/pull/30719#issuecomment-682661817 for the future plans of this PR. As soon as the conversion into a custom component is ready, I'll post a link here for those who want to give it a shot.

superm1 commented:
@rytilahti did you end up with any luck in the creation of custom component?  I just picked up some KL125P2's that don't work with the existing component, but do work with `python-kasa`.  Happy to be a guinea pig if you're looking for one.

rytilahti commented:
Hi, unfortunately I haven't had time to clean up the code base properly and I'm currently occupied with more important projects, so this is on hold :-(

superm1 commented:
> Hi, unfortunately I haven't had time to clean up the code base properly and I'm currently occupied with more important projects, so this is on hold :-(

No worries.  My main goal is getting those bulbs working now, and it looks like my other PR will be accepted and we're in good shape.  Whenever you get cycles to come back, happy to help with it.

mike391 commented:
From looking around online, people in the HASS forums have reported success by having a script run on a PC that pings all their kasa devices. Apparently they require a ping every so often or they will fail to respond to the ARP requests HASS sends out when checking their availability.  Would this be a relatively simple fix to implement?

saicrazyfire commented:
@mike391 can you link to one of those posts?
